### PR TITLE
Udisks2 support

### LINKED ISSUE #3264: udisks2 support
_From [Nickollai](https://code.google.com/u/Nickollai/) on November 05, 2012 20:26:43_

https://bugs.gentoo.org/show_bug.cgi?id=439644 Current clementine docs do not tell anything about supported udisks versions. Both wiki and git repo. Please add and verify udisks2 support and document it at https://code.google.com/p/clementine-player/wiki/PortableDevices .

_Original issue: http://code.google.com/p/clementine-player/issues/detail?id=3263_

### PR DESCRIPTION
**Pull request by Chemrat**
Should fix #3264

### CODE REVIEW

File: src/devices/udisks2lister.h [hunk 1]
```diff
@@ -0,0 +1,93 @@
+#pragma once
+
+#include <memory>
+
+#include <QReadWriteLock>
```
Comment (hatstand) [inline src/devices/udisks2lister.h]: alpha sort headers

File: src/devices/udisks2lister.h [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+  void DBusInterfaceRemoved(const QDBusObjectPath &path, const QStringList &ifaces);
+  void JobCompleted(bool success, const QString &message);
+
+private:
+  bool isPendingJob(const QDBusObjectPath &jobPath);
+  void RemoveDevice(const QDBusObjectPath &devicePath);
+  QList<QDBusObjectPath> GetMountedPartitionsFromDBusArgument(const QDBusArgument &input);
+
+  struct Udisks2Job {
+    bool isMount = true;
```
Comment (hatstand) [inline src/devices/udisks2lister.h]: `is_mount`

File: src/devices/udisks2lister.h [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+  void HandleFinishedMountJob(const Udisks2Lister::PartitionData &partitionData);
+  void HandleFinishedUnmountJob(const Udisks2Lister::PartitionData &partitionData, const QDBusObjectPath &mountedObject);
+
+  QReadWriteLock device_data_lock_;
+  QMap<QString, PartitionData> device_data_;
+
+private:
+  std::unique_ptr<OrgFreedesktopDBusObjectManagerInterface> udisks2_interface_;
+
+  static const QString udisks2service_;
```
Comment (hatstand) [inline src/devices/udisks2lister.h]: We avoid non-POD static initialisation so just use a `char*` or `char[]`

File: src/devices/udisks2lister.cpp [hunk 4]
```diff
@@ -0,0 +1,369 @@
...
+#include "dbus/udisks2job.h"
+
+const QString Udisks2Lister::udisks2service_ = "org.freedesktop.UDisks2";
+
+Udisks2Lister::Udisks2Lister() {
+
+}
+
+Udisks2Lister::~Udisks2Lister() {
+  qLog(Debug) << __PRETTY_FUNCTION__;
```
Comment (hatstand) [inline src/devices/udisks2lister.cpp]: Remove

File: src/devices/udisks2lister.cpp [hunk 5]
```diff
@@ -0,0 +1,369 @@
...
+
+    std::shared_ptr<OrgFreedesktopUDisks2JobInterface> job = std::make_shared<OrgFreedesktopUDisks2JobInterface>(
+          udisks2service_,
+          path.path(),
+          QDBusConnection::systemBus());
+
+    if (!job->isValid())
+      continue;
+
+    bool isMountJob = false;
```
Comment (hatstand) [inline src/devices/udisks2lister.cpp]: `is_mount_job`

**Comment by hatstand** [comment]
Please run `make format` too

File: src/devices/udisks2lister.cpp [hunk 6]
```diff
@@ -0,0 +1,366 @@
...
+void Udisks2Lister::UnmountDevice(const QString& id) {
+  QReadLocker locker(&device_data_lock_);
+  if (!device_data_.contains(id)) return;
+
+  OrgFreedesktopUDisks2FilesystemInterface filesystem(
+      udisks2_service_, device_data_[id].dbus_path,
+      QDBusConnection::systemBus());
+
+  if (filesystem.isValid()) {
+    auto unmountResult = filesystem.Unmount(QVariantMap());
```
Comment (hatstand) [inline src/devices/udisks2lister.cpp]: `unmount_result`

File: src/devices/udisks2lister.cpp [hunk 7]
```diff
@@ -0,0 +1,366 @@
...
+                    << unmountResult.error();
+      return;
+    }
+
+    OrgFreedesktopUDisks2DriveInterface drive(udisks2_service_,
+                                              device_data_[id].dbus_drive_path,
+                                              QDBusConnection::systemBus());
+
+    if (drive.isValid()) {
+      auto ejectResult = drive.Eject(QVariantMap());
```
Comment (hatstand) [inline src/devices/udisks2lister.cpp]: `eject_result`

File: src/devices/udisks2lister.cpp [hunk 8]
```diff
@@ -0,0 +1,366 @@
...
+
+  if (!reply.isValid()) {
+    qLog(Warning) << "Error enumerating udisks2 devices:"
+                  << reply.error().name() << reply.error().message();
+    udisks2_interface_.reset();
+    return;
+  }
+
+  for (const QDBusObjectPath& path : reply.value().keys()) {
+    auto partitionData = ReadPartitionData(path);
```
Comment (hatstand) [inline src/devices/udisks2lister.cpp]: `partition_data`

File: src/devices/udisks2lister.cpp [hunk 9]
```diff
@@ -0,0 +1,366 @@
...
+    if (interface.key() != "org.freedesktop.UDisks2.Job") continue;
+
+    std::shared_ptr<OrgFreedesktopUDisks2JobInterface> job =
+        std::make_shared<OrgFreedesktopUDisks2JobInterface>(
+            udisks2_service_, path.path(), QDBusConnection::systemBus());
+
+    if (!job->isValid()) continue;
+
+    bool is_mount_job = false;
+    if (job->operation() == "filesystem-mount")
```
Comment (hatstand) [inline src/devices/udisks2lister.cpp]: Add `{}`

File: src/devices/udisks2lister.cpp [hunk 10]
```diff
@@ -0,0 +1,366 @@
...
+      is_mount_job = true;
+    else if (job->operation() == "filesystem-unmount")
+      is_mount_job = false;
+    else
+      continue;
+
+    auto mountedPartitions = job->objects();
```
Comment (hatstand) [inline src/devices/udisks2lister.cpp]: `mounted_partitions`

File: src/devices/udisks2lister.cpp [hunk 11]
```diff
@@ -0,0 +1,366 @@
...
+  QDBusObjectPath jobPath(job->path());
+
+  if (!job->isValid() || !success || !mounting_jobs_.contains(jobPath)) return;
+
+  qLog(Debug) << "Pending Job Completed | Path = " << job->path()
+              << " | Mount? = " << mounting_jobs_[jobPath].is_mount
+              << " | Success = " << success;
+
+  for (const auto& mountedObject : mounting_jobs_[jobPath].mounted_partitions) {
+    auto partitionData = ReadPartitionData(mountedObject);
```
Comment (hatstand) [inline src/devices/udisks2lister.cpp]: `partition_data`

**Comment by hatstand** [comment]
Almost there; please add a comment when you update it otherwise I don't get notified.

**Comment by Chemrat** [comment]
Fixed rest of the variable names

Just a side note, clang seems to have an utility for formatting variable names (clang-tidy) but it needs some minor build system integration

**Comment by hatstand** [comment]
I think the formatter is here: https://github.com/davidsansome/clang-in-the-cloud

**Pull request merged**